homeassistant/components/vasttrafik/config_flow.py

```diff
@@ -0,0 +1,351 @@
+        return [], "api_error"
+    except Exception:
+        _LOGGER.exception("Unexpected error in search_stations")
+        return [], "api_error"
+
+    stations: list[SelectOptionDict] = []
+    for result in results:
+        name = result.get("name", "")
+        gid = result.get("gid", "")
+        if name and gid:
+            stations.append({"value": name, "label": f"{name} ({gid})"})
+
+    if stations:
+        return stations, None
+    return [], "no_stations_found"
+
+
+async def validate_api_credentials(
+    hass: HomeAssistant, data: dict[str, Any]
+) -> dict[str, str] | None:
+    """Validate that the user input allows us to connect.
+
+    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
+
+    Returns errors if any
+    """
+    planner = await hass.async_add_executor_job(
+        vasttrafik.JournyPlanner, data[CONF_KEY], data[CONF_SECRET]
+    )
+
+    try:
+        await hass.async_add_executor_job(planner.location_name, "Centralstationen")
+    except vasttrafik.Error as err:
+        _LOGGER.error("Failed to validate Västtrafik credentials: %s", err)
+        if (
+            hasattr(err, "response")
+            and err.response
+            and err.response.status_code in (401, 403)
+        ):
+            return {"base": "invalid_auth"}
+        return {"base": "cannot_connect"}
+    except Exception as err:  # noqa: BLE001
+        _LOGGER.error("Unexpected error validating Västtrafik credentials: %s", err)
+        return {"base": "unknown"}
+
+    return None
+
+
+class VasttrafikConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Västtrafik."""
+
+    VERSION = 1
+    MINOR_VERSION = 1
+
+    @classmethod
+    @callback
+    def async_get_supported_subentry_types(
+        cls, config_entry: ConfigEntry
+    ) -> dict[str, type[ConfigSubentryFlow]]:
+        """Get supported subentry types."""
+        return {"departure_board": VasttrafikSubentryFlow}
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the initial step."""
+
+        errors: dict[str, str] | None = {}
+        if user_input is not None:
+            if not (errors := await validate_api_credentials(self.hass, user_input)):
+                return self.async_create_entry(title="Västtrafik", data=user_input)
+
+        return self.async_show_form(
+            step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors or {}
+        )
+
+    async def async_step_reconfigure(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle reconfiguration of API credentials."""
+        entry = self._get_reconfigure_entry()
+
+        errors: dict[str, str] | None = {}
+        if user_input is not None:
+            if not (errors := await validate_api_credentials(self.hass, user_input)):
+                return self.async_update_reload_and_abort(
```

Comment:
```suggestion
                return self.async_update_and_abort(
```
This should not reload since you have an update listener which does it